packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ClientProvider.cs

```diff
@@ -31,18 +63,203 @@ public ClientProvider(InputClient inputClient)
+                    parameter.Type,
+                    "_" + parameter.Name.ToVariableName());
+                fields.Add(field);
+                ctorParametersNamesToFields[parameter.Name] = field;
+            }
+
+            _ctorParametersNamesToFields = ctorParametersNamesToFields;
+
+            return [.. fields];
         }
 
         protected override ConstructorProvider[] BuildConstructors()
         {
-            return
-            [
-                new ConstructorProvider(
-                    new ConstructorSignature(Type, $"{_inputClient.Description}", MethodSignatureModifiers.Public, []),
-                    new MethodBodyStatement[] { PipelineField.Assign(ClientPipelineSnippets.Create()).Terminate() },
-                    this)
-            ];
+            var requiredParameters = _apiKeyAuthField != null ? [ScmKnownParameters.ApiKeyAuth] : new List<ParameterProvider>();
+            var optionalParameters = new List<ParameterProvider>();
+            var orderedParams = new List<ParameterProvider>(_clientParameters.Count);
+            ParameterProvider? endpointParam = null;
+
+            foreach (var parameter in _clientParameters)
+            {
+                if (parameter.Name == KnownParameters.Endpoint.Name)
+                {
+                    endpointParam = parameter;
+                    orderedParams.Add(endpointParam);
+                }
+                else if (parameter.DefaultValue == null && parameter.Validation != ParameterValidationType.None)
+                {
+                    requiredParameters.Add(parameter);
+                }
+                else
+                {
+                    optionalParameters.Add(parameter);
+                }
+            }
+
+            // order the constructor parameters as (endpoint, requiredParameters, optionalParameters).
+            orderedParams.AddRange(requiredParameters);
+            optionalParameters.Add(_clientOptionsParameter);
+            orderedParams.AddRange(optionalParameters);
+
+            var primaryConstructor = new ConstructorProvider(
+                new ConstructorSignature(Type, _publicCtorDescription, MethodSignatureModifiers.Public, orderedParams),
+                BuildPrimaryConstructorBody(orderedParams),
+                this);
+            var secondaryConstructor = BuildSecondaryConstructor(requiredParameters, orderedParams, endpointParam);
+
+            if (requiredParameters.Count > 0 || _apiKeyAuthField != null)
+            {
+                return [ConstructorProvider.BuildMockingConstructor(this), secondaryConstructor, primaryConstructor];
+            }
+            else
+            {
+                return [secondaryConstructor, primaryConstructor];
+            }
+        }
+
+        private MethodBodyStatement[] BuildPrimaryConstructorBody(IReadOnlyList<ParameterProvider> orderedParameters)
+        {
+            MethodBodyStatement? clientOptions = null;
```

Comment:
we can reduce the complexity here by simply adding the clientoptions lines first then looping on the rest and skipping the clientoptions since we know this will always exist.